README.md
Seeds.js - A Full Stack JavaScript Framework
=====

[![Join the chat at https://gitter.im/terminalvelocity/seeds.js](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/terminalvelocity/seeds.js?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Stories in Ready](https://badge.waffle.io/terminalvelocity/seeds.js.png?label=ready&title=Ready)](https://waffle.io/terminalvelocity/seeds.js) ![Downloads on NPM](http://img.shields.io/npm/dm/seeds.svg?style=flat-square)

[![](https://nodei.co/npm/seeds.png?&downloadRank=true)](https://nodei.co/npm/seeds.png?downloads=true&downloadRank=true)

Seeds is an acronym that stands for **S**ails **E**mber **E**mber-**D**ata **S**emantic-UI. It is designed for Rapid Application Prototyping.

 [View full documentation.](http://docs.seedsjs.com)

```
                                                         .
                                              .         ;
                 .              .              ;%     ;;
                   ,           ,                :;%  %;
                    :         ;                   :;%;'     .,
           ,.        %;     %;            ;        %;'    ,;
             ;       ;%;  %%;        ,     %;    ;%;    ,%'
              %;       %;%;      ,  ;       %;  ;%;   ,%;'
               ;%;      %;        ;%;        % ;%;  ,%;'
                `%;.     ;%;     %;'         `;%%;.%;'
                 `:;%.    ;%%. %@;        %; ;@%;%'
                    `:%;.  :;bd%;          %;@%;'
                      `@%:.  :;%.         ;@@%;'
                        `@%.  `;@%.      ;@@%;
                          `@%%. `@%%    ;@@%;
                            ;@%. :@%%  %@@%;
                              %@bd%%%bd%%:;
                                #@%%%%%:;;
                                %@@%%%::;
                                %@@@%(o);  . '
                                %@@@o%;:(.,'
                            `.. %@@@o%::;
                               `)@@@o%::;
                                %@@(o)::;
                               .%@@@@%::;
                               ;%@@@@%::;.
                              ;%@@@@%%:;;;.
                          ...;%@@@@@%%:;;;;,..
                    Plant some seeds, watch em grow
                        You reap what you sow
```

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Seeds.js**

- [Installation](#installation)
- [Usage](#usage)
  - [Generates a new project with the supplied name](#generates-a-new-project-with-the-supplied-name)
  - [Run API Server at 1776 and Frontend at 4200](#run-api-server-at-1776-and-frontend-at-4200)
  - [Generates boilerplate of <kind> named [name]](#generates-boilerplate-of-kind-named-name)
    - [Currently only scaffold generator is available.](#currently-only-scaffold-generator-is-available)
- [Note](#note)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Installation
`npm install -g seeds`

## Usage

### Generates a new project with the supplied name
```
$ seeds new [name]
```

### Run API Server at 1776 and Frontend at 4200
```
$ seeds serve
```

### Generates boilerplate of <kind> named [name]
```
$ seeds generate <kind> [name] <attrs:type>
```

#### Currently only scaffold generator is available.

```
$ seeds generate scaffold user name:string email:string age:number birthday:date newsletter:boolean
```


## Note

Seeds.js uses tagged releases. The latest tagged version is the latest version available on NPM. Master is dev master branch.


CHANGELOG.md
# Changelog

## v2.3.1 | 2016-04-06
* Added link to docs to npm
* runs multiple ember servers. WIP

## v2.3.0 | 2016-03-09
* fix error message for db
* lib/template
* update seeds-backend package
* adds db commands to seeds

## v2.2.2 | 2016-03-09
* Update sails version number in the version command's output

## v2.2.1 | 2016-03-09
* bump seeds-frontend and backend versions
* bumps to latest seeds-frontend package
* update seeds-backend package and sails dep
* remove resource generator from generate.es6
* Fixes model generator to use scaffold-model so that it always uses top level model folder with pods structure. Removes resource generator as I don't like what it creates and do not want to override it
* Add download rank badge

## v2.2.0 | 2016-03-09
* adds --cache-min to install command for api/frontend. Reduces new command time by ~50%

## v2.1.5 | 2016-03-09
* reset version number
* removed extraneous test.txt file
* Updates version numbers in version command and adds version to seedsrc file on seeds new
* Update Ember to 2.4.1, updates generate-seeds-frontend version to latest version with Ember 2.4.1 templates, remove hardcoded bower deps and replace with a bower install during ember install.
* Adds lib to git

## v2.1.4 | 2016-03-06
* changes templates package.json file to use tilde (minor level matcher) for generators

## v2.1.3 | 2016-02-25


## v2.1.2 | 2016-02-25
* Modified help command. Also, adds alias for help & build commands
* Adds build command. Build creates a timestamped directory in the builds dir, copies api server, then runs ember build command with output-dir set to timestamped api server's asset dir. Adds mkdirp as dep. Updates default seeds .gitignore file to ignore the builds directory
* Refactored commands.es6 to automatically export all files found in the commands dir as CLI commands

## v2.1.1 | 2016-02-15
* Adds resource generator to seeds. seeds g resource modelName attrs:type

## v2.1.0 | 2016-02-11
* Adds model generator to seeds. Creates API model and controller as well as Ember Model.

## v2.0.7 | 2016-02-09
* Adds init command to install API and Frontend dependencies for a seeds project

## v2.0.6 | 2015-09-23
* add lib to .gitignore
* Adds gitignore to templates for seeds new command. Changed .seedsrc to seedsrc in templates so babel will see it and changed new command to copy over the unhidden version into a hidden version, does same for .gitignore. Removes Lib and fixes #62

## v2.0.5 | 2015-09-17
* Cleans up commands.es6 to use an array of command names to export

## v2.0.4 | 2015-09-16
* Adds versions table and help output
* Fixes npm run build and watch to only compile .es6 files and copy over other files to lib from src. Currently does not copy .seedsrc file. May change the .seedsrc file creation process

## v2.0.3 | 2015-09-14
* Reverts seeds version.
* Ember 2.0 compatible.
* Ember 2.0 compatible.
* preps for multiple front and backend apps as well as utilizing rails as a backend

## v2.0.2 | 2015-09-11
* Reverts bin/seeds. Want cli to work with version of node older than 4.
* move banner.es6 to lib to skip transpile. Babel does not preserve spacing of comments, also removes sails/ember from deps since it's not needed
* remove accidental tests.js from project root, update package.json test script and fixed typo in tests.es6
* Add dummy to npm ignore
* Template strings in node4 doesn't work 100% as expected. Reverted to old version of banner using multiline
* Cleans up repo. Moves pre-transpiled files to src folder, merge lib/settings and lib/helpers folder, updates package.json scripts to new setup
* Updates bin/seeds to es6
* removes ./node_modules/.bin from build/watch since it's not required when using npm run <command>, via @IanVS

## v2.0.1 | 2015-06-24
* fixes output of install command to show only which directory inside the current directory we are installing to instead of the absolute path
* adds test for install alias
* adds install command + alias 'i'

## v2.0.0 | 2015-06-19
* changes release hooks babel call from build-es5 to build
* adds *.es6 to npmignore
* Removes comments from new command
* removes extraneous templates, cleans up new command
* Uses sails generate seeds-backend instead of sails new
* adds test to ensure debug is false before pushing
* Adds tests to npm run watch transpile command
* debug: false by default
* Finishes up es6 transpile. all commands work on node 0.12.4
* Adds babel as a devdep
* sets .seedsrc up for multiple front end apps
* rm npm-debug.log.randomstring
* es6 -> es5
* readds templates to eslintrc
* Adds watch/build to package.json, transpiles es6 to es5 and works in node v0.12.4
* es6-transpile wip
* debug flag false by default
* Upgrades soil-cli version
* table of contents update
* Updates README
* Adds info comment about model types
* Adds package.json to templates
* Updates links in package.json
* add npm run watch command to have tests run automatically on filesave
* Adds test for help to be default command
* adds tests for aliases
* Adds aliases/flags class. Called in the constructor to parse aliases and flags. Flags -v, --version, -h, and --help will supercede all other commands
* Updates test to reflect change to conver models file. Camelcased instead of dashed
* Refactors generate command to use the updated convertModels command. Data types are now properly converted from for sails and ember. Also normalized all the directory calls so it should now work on windows
* refactor generate to check for existance of .seedsrc file instead of checking for specific dir names
* Adds tests for converting model datatypes
* Fixed eslint error 'guard-for-in', removes convert-models from eslintignore
* Refactors the convert-models function
* Adds mocha: true to eslint, ads a couple tests for seedsrc defaults
* moves debug flag to top of constructor
* debug flag is based on config
* use the seeds config for file names, changes all the references to paths in lib/cli to use path.join

lib/commands/new.js
/*eslint no-undefined:0*/
/*eslint quote-props:0*/
/*eslint quotes:0*/

'use strict';

var bower = require('bower');
var join = require('path').join;
var copy = require('fs-extra').copySync;
var child = require('child_process');
var spawnSync = child.spawnSync;
var fs = require('fs');

module.exports = function (cli) {
  var appName = cli.args[1];
  var appDir = join(cli.cwd, appName);
  var apiDir = join(appDir, cli.apiName);
  var feDir = join(appDir, cli.feName);
  var templatesPath = join(cli.includedBasepath, 'lib', 'templates');

  var createAppFolder = function createAppFolder() {
    cli.debug('createAppFolder', appName);

    if (!appName) {
      cli.error('You must supply a name for your Seeds.js Application.');
      cli.exit(1);
    }

    cli.ui('Generating a new Seed named'.green(), appName.white() + '...'.green());
    cli.mkdir(feDir);
    return cli.mkdir(apiDir);
  };

  var bootstrapApp = function bootstrapApp() {
    cli.debug('bootstrapApp');
    var rc = {
      "apps": { "sails": [{ "name": "api", "port": 1776 }], "ember": [{ "name": "frontend", "port": 4200 }] },
      "debug": false,
      "version": cli['package'].version
    };

    copy(join(templatesPath, 'gitignore'), join(appDir, '.gitignore'));
    fs.writeFileSync(join(appDir, '.seedsrc'), JSON.stringify(rc, null, '\t'));
    copy(join(templatesPath, 'package.json'), join(appDir, 'package.json'), { clobber: true });

    spawnSync('npm', ['install'], { cwd: join(cli.cwd, appName) });
    return;
  };

  var setupSails = function setupSails() {
    cli.debug('setupSails');

    spawnSync(join(appDir, 'node_modules', '.bin', 'sails'), ['generate', 'seeds-backend'], { cwd: apiDir });
    spawnSync('npm', ['--cache-min', '9999999', 'install'], { cwd: apiDir });
    return;
  };

  var setupEmber = function setupEmber() {
    cli.debug('setupEmber');

    spawnSync(join(appDir, 'node_modules', '.bin', 'sails'), ['generate', 'seeds-frontend'], { cwd: feDir });
    spawnSync('npm', ['--cache-min', '9999999', 'install'], { cwd: feDir });
    bower.commands.install(undefined, undefined, { cwd: feDir });
    return;
  };

  var wrapUp = function wrapUp() {
    cli.debug('wrapUp');

    cli.banner(join(cli.includedBasepath, 'lib', 'helpers', 'banner'));
    return;
  };

  createAppFolder().then(function () {
    bootstrapApp();
    setupSails();
    setupEmber();
    wrapUp();
  });
};